hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/SaslUtil.java

```diff
@@ -0,0 +1,195 @@
+   * @return CipherOption negotiated cipher option
+   */
+  public static CipherOption negotiateCipherOption(Configuration conf,
+      List<CipherOption> options) throws IOException {
+    // Negotiate cipher suites if configured.  Currently, the only supported
+    // cipher suite is AES/CTR/NoPadding, but the protocol allows multiple
+    // values for future expansion.
+    String cipherSuites = conf.get(HADOOP_SECURITY_CRYPTO_CIPHER_SUITE_KEY);
+    if (cipherSuites == null || cipherSuites.isEmpty()) {
+      return null;
+    }
+    if (!cipherSuites.equals(CipherSuite.AES_CTR_NOPADDING.getName())) {
+      throw new IOException(String.format("Invalid cipher suite, %s=%s",
+          HADOOP_SECURITY_CRYPTO_CIPHER_SUITE_KEY, cipherSuites));
+    }
+    if (options != null) {
+      for (CipherOption option : options) {
+        CipherSuite suite = option.getCipherSuite();
+        if (suite == CipherSuite.AES_CTR_NOPADDING) {
+          int keyLen = conf.getInt(
+              HADOOP_SECURITY_CRYPTO_CIPHER_KEY_BITLENGTH_KEY,
+              HADOOP_SECURITY_CRYPTO_CIPHER_KEY_BITLENGTH_DEFAULT) / 8;
+          CryptoCodec codec = CryptoCodec.getInstance(conf, suite);
+          byte[] inKey = new byte[keyLen];
+          byte[] inIv = new byte[suite.getAlgorithmBlockSize()];
+          byte[] outKey = new byte[keyLen];
+          byte[] outIv = new byte[suite.getAlgorithmBlockSize()];
+          assert codec != null;
+          codec.generateSecureRandom(inKey);
+          codec.generateSecureRandom(inIv);
+          codec.generateSecureRandom(outKey);
+          codec.generateSecureRandom(outIv);
+          return new CipherOption(suite, inKey, inIv, outKey, outIv);
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Encrypt the key and iv of the negotiated cipher option.
+   *
+   * @param option negotiated cipher option
+   * @param saslServer SASL server
+   * @return CipherOption negotiated cipher option which contains the
+   * encrypted key and iv
+   * @throws IOException for any error
+   */
+  public static CipherOption wrap(CipherOption option, SaslServer saslServer)
+      throws IOException {
+    if (option != null) {
+      byte[] inKey = option.getInKey();
+      if (inKey != null) {
+        inKey = saslServer.wrap(inKey, 0, inKey.length);
+      }
+      byte[] outKey = option.getOutKey();
+      if (outKey != null) {
+        outKey = saslServer.wrap(outKey, 0, outKey.length);
+      }
+      return new CipherOption(option.getCipherSuite(), inKey, option.getInIv(),
+          outKey, option.getOutIv());
+    }
+
+    return null;
+  }
+
+  /**
+   * Decrypt the key and iv of the negotiated cipher option.
+   *
+   * @param option negotiated cipher option
+   * @param saslClient SASL client
+   * @return CipherOption negotiated cipher option which contains the
+   * decrypted key and iv
+   * @throws IOException for any error
+   */
+  public static CipherOption unwrap(CipherOption option, SaslClient saslClient)
```

Comment:
The comment says "Decrypt the key and iv of the negotiated cipher option", but iv doesn't be decrypted.